### PR TITLE
File operator

### PR DESCRIPTION
**Pull request by kleintom**
Add a 'file:' operator for matching on file names only. Fixes bug 1249325.

### CODE REVIEW

**Comment by kleintom** [comment]
(There was some mention at one point of removing the 'ext:' operator once 'file:' was implemented:
https://github.com/erikrose/dxr/blob/query-parser/docs/queries.rst
I'm not sure we'd want to do that at this point (and I would do it in a different PR in any case), but bringing it up here just in case.)

File: dxr/plugins/core.py [hunk 1]
```diff
@@ -308,6 +301,26 @@ def filter(self):
             raise BadTerm('Path globs need at least 3 literal characters in a row '
                           'for speed.')
```
Comment (pelmers) [inline dxr/plugins/core.py]: 2 blank lines

File: dxr/plugins/core.py [hunk 2]
```diff
@@ -21,11 +22,11 @@
                             NoTrigrams, PythonRegexVisitor)
 from dxr.utils import glob_to_regex
 
-__all__ = ['mappings', 'analyzers', 'TextFilter', 'PathFilter', 'ExtFilter',
-           'RegexpFilter', 'IdFilter', 'RefFilter']
+__all__ = ['mappings', 'analyzers', 'TextFilter', 'PathFilter', 'FileFilter',
```
Comment (pelmers) [inline dxr/plugins/core.py]: Here it's FileFilter, but later it's defined FilenameFilter

File: dxr/plugins/core.py [hunk 3]
```diff
@@ -308,6 +301,26 @@ def filter(self):
...
+                         '<code>?</code>, and <code>[...]</code> act '
+                         'as shell wildcards.')
+
+    @negatable
+    def filter(self):
+        glob = self._term['arg']
+        try:
+            return es_regex_filter(
+                regex_grammar.parse(glob_to_regex(glob)),
+                'file_name',
```
Comment (pelmers) [inline dxr/plugins/core.py]: Here's the only difference with PathFilter; would it be reasonable to refactor into subclasses?
Perhaps inherit of something like PathSegmentFilter.

File: dxr/es.py [hunk 4]
```diff
@@ -13,6 +13,12 @@
 }
 
 
+UNANALYZED_STRING = {
```
Comment (pelmers) [inline dxr/es.py]: nice refactoring 🏆

File: dxr/plugins/core.py [hunk 5]
```diff
@@ -2,15 +2,16 @@
 
 from base64 import b64encode
 from itertools import chain
-from os.path import relpath, splitext, islink, realpath
+from os.path import relpath, splitext, islink, realpath, basename
```
Comment (pelmers) [inline dxr/plugins/core.py]: while we touch this line, I note that the file never calls islink

File: dxr/plugins/core.py (hunk 2 above)
Comment (kleintom) [inline dxr/plugins/core.py]: > Here it's FileFilter, but later it's defined FilenameFilter

Nice catch!

File: dxr/plugins/core.py (hunk 3 above)
Comment (kleintom) [inline dxr/plugins/core.py]: I'm still on the fence... the `name` and `description` vars also differ, as does the `BadTerm` text.  Seems like you'd be factoring out basically just the call to `es_regex_filter` (with some parameter for the mapping name)...  I'm leaning toward saying it's not worth it - what do you think?

**Comment by kleintom** [comment]
@pelmers I wound up factoring out a _PathSegmentFilterBase - can you take a look?

**Comment by pelmers** [comment]
Looks great, r+

**Comment by kleintom** [comment]
Thanks for the reviews Peter!

**Pull request merged**